Fix SQLiteDatabase::turnOnIncrementalAutoVacuum()

bnham opened the pull request:
#### a4485084ec04ac3b43b1eb7c0ae537b7bcc5401f
<pre>
Fix SQLiteDatabase::turnOnIncrementalAutoVacuum()
<a href="https://bugs.webkit.org/show_bug.cgi?id=259284">https://bugs.webkit.org/show_bug.cgi?id=259284</a>
rdar://112475480

Reviewed by Chris Dumez.

PulkoMandy from the Haiku port found that turnOnIncrementalAutoVacuum is broken again on trunk due
to 258673@main. (It was previously fixed in 241334@main.)

The issue is that we need to call lastError() immediately after the call to columnInt. What we were
doing was calling it immediatley after the destructor for the PRAGMA auto_vacuum statement. So
instead of checking the return code for the statement running, we were checking the return code for
finalizing the statement.

* Source/WebCore/platform/sql/SQLiteDatabase.cpp:
(WebCore::SQLiteDatabase::turnOnIncrementalAutoVacuum):

Canonical link: <a href="https://commits.webkit.org/266139@main">https://commits.webkit.org/266139@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ae582fb820fbc1c59ea066b855a20864de6c4e11

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/13016 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/13334 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/13665 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/14754 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/12410 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/13084 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/15840 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/13360 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/15104 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/13183 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/13878 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/11002 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/15207 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/11166 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/11757 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/18824 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/12241 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/11923 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/15137 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/12403 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/10288 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/11670 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3188 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/15988 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/12246 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/platform/sql/SQLiteDatabase.cpp:
```diff
@@ -673,16 +673,16 @@ bool SQLiteDatabase::turnOnIncrementalAutoVacuum()
+        // auto_vacuum flag and try to set it to INCREMENTAL the next time we open this
+        // database. If the error is not SQLITE_BUSY, then we probably ran into a more
+        // serious problem and should return false (to log an error message).
+        if (lastError() != SQLITE_ROW)
```
Can you please add a comment explaining that the call to lastError() needs to occur while the transaction is still active/alive. Hopefully, it will discourage somebody from breaking this again in the future.

cdumez reviewed (approved):
LGTM

bnham committed:
Fix SQLiteDatabase::turnOnIncrementalAutoVacuum()
https://bugs.webkit.org/show_bug.cgi?id=259284
rdar://112475480

Reviewed by Chris Dumez.

PulkoMandy from the Haiku port found that turnOnIncrementalAutoVacuum is broken again on trunk due
to 258673@main. (It was previously fixed in 241334@main.)

The issue is that we need to call lastError() immediately after the call to columnInt. What we were
doing was calling it immediatley after the destructor for the PRAGMA auto_vacuum statement. So
instead of checking the return code for the statement running, we were checking the return code for
finalizing the statement.

* Source/WebCore/platform/sql/SQLiteDatabase.cpp:
(WebCore::SQLiteDatabase::turnOnIncrementalAutoVacuum):

Canonical link: https://commits.webkit.org/266139@main
